0005-on-chain-voting-protocol-proposal.md

```diff
@@ -0,0 +1,195 @@
+
+### Deck vote init
+
+To enable voting on the deck, deck issuer has to make first blank vote_init and send it to deck_vote_tag.
+This transaction is called deck_vote_init.
+This transaction has to originate from the deck spawn address.
+
+
+## Vote casting
+
+Vote is casted by sending sending transaction from address which has positive card balance to address derived from sha256(vote_init + #OPTION).
+Vote_cast receiver address is calculated deterministically from vote_init and vote option index.
+
+For example, Bob is voting for on choices offered by vote_init in example above. Bob wants to vote for "trump" option, which is third option of the vote.
+Bob calculates vote_cast address in Python using pypeerassets:
+
+```
+import pypeerassets as pa
+from hashlib import sha256
+from binascii import unhexlify
+
+vote_init_txid = unhexlify("b461051e1d9147770ec9d4c254bdd68ff5d203a86b97c09d00955fb3f088cab3")
+option_index = vote_init["choices"].index("trump") # 2
+
+vote_cast_privkey = sha256(vote_init_txid + bytes(option_index)).hexdigest()
+
+vote_cast_address = pa.Kutil(network="tppc", privkey=vote_cast_privkey).address
+
+assert vote_cast_address == "msCpEmB3m2sARUikdZSCgyoCokmXykD1de"
+
+```
+
+All clients can deterministically calculate this address and check transactions paying to it to verify if they are legit vote_casts.
+
+
+## Vote counting
+
+Simplest form of vote counting is vote per transaction, with filtering of double votes and invalid votes.
+The only requisite of valid vote is positive card balance with this schema. However this schema is easily gamed, so two more complex schemas are proposed bellow.
+
+### Weighting votes with PA card balance
+
+This schema allows weighting vote with PeerAssets card balance. This allows equating casted vote with stake in the deck.
+Simplest form of card balance weighted voting is using card balance when summing up votes.
+
+Example:
+
+> Voting options are "A" and "B".
+
+    Alice_card_balance = 102.0
+    Bob_card_balance = 80.0
+    David_card_balance = 27.0
+
+    Alice votes for option "A", while Bob and David vote "B".
+    Option A weight is 102
+    Option B weight is 107
+
+    Option B wins.
+
+
+#### Weighting vote with PA card balance with "lookback period"
+
+This scheme prevents actor to purchase cards during or just before the vote and them use them to cast weighted vote and then simply sell them.
+Lookback period schema can be applied to example above, where balance of cards is checked with lookback period of 10 blocks.
+
+> Voting options are "A" and "B", cut-off time is 10 blocks
+
+    Alice_card_balance = 40.0
+    Bob_card_balance = 80.0
+    David_card_balance = 27.0
+
+    Alice votes for option "A", while Bob and David vote "B".
+    Option A weight is 40
+    Option B weight is 107
+
+    Option B wins.
+
+
+In this example Alice tried to influence the vote by buying more cards after the vote begun, however those cards are not counted due to lookback period.
```

Comment:
I don't understand the lookback period, why not just conlude the vote at a predifinded block height?